Java: Fix WildcardTypeAccess.hasNoBound() not considering ArrayTypeAccess

Marcono1234 opened the pull request:
- Fixes #4005
- Adds `WildcardTypeAccess.getBound()`
- Improves `WildcardTypeAccess.toString()`
- Extends tests

aschackmull commented on java/ql/test/library-tests/locations/WildcardLocations.ql:
```diff
@@ -1,4 +1,30 @@
+  else
+    // Also makes sure that getBound() is working correctly (and has at most
+    // one result)
+    exists(Expr bound | bound = wta.getBound() and not wta.getBound() != bound |
```
```suggestion
    exists(Expr bound | bound = wta.getBound() |
```
It's better to test the not-having-multiple-values by potentially observing those values rather than observing nothing.

Marcono1234 replied:
Though then the test would also succeed when `getBound()` would have completely unrelated results since here in the test its result is bound below with `getUpperBound()` / `getLowerBound()`.

aschackmull commented on java/ql/test/library-tests/locations/WildcardLocations.ql:
```diff
@@ -1,4 +1,30 @@
+class WildcardTypeAccessTest extends InlineExpectationsTest {
+  WildcardTypeAccessTest() { this = "WildcardTypeAccessTest" }
+
+  override string getARelevantTag() { result = ["wildcardTypeAccess"] }
```
Simplify:
```suggestion
  override string getARelevantTag() { result = "wildcardTypeAccess" }
```

aschackmull commented on java/ql/src/semmle/code/java/Expr.qll:
```diff
@@ -1742,11 +1742,27 @@ class WildcardTypeAccess extends Expr, @wildcardtypeaccess {
   Expr getLowerBound() { result.isNthChildOf(this, 1) }
 
+  /**
+   * Gets the bound of this wildcard type access, if any, i.e. either the upper
```
Don't use abbreviations in qldoc.
```suggestion
   * Gets the bound of this wildcard type access, if any, that is, either the upper
```

Marcono1234 replied:
Made this a separate sentence ("... any. That is, ...") to be consistent with other comments and to not have so many commas.

aschackmull reviewed (commented):
Some inline comments, otherwise LGTM.

aschackmull commented:
```
Some tests failed!
    ql/java/ql/test/library-tests/typeaccesses/ArrayTypeAccesses.ql: Extractor error (Extraction failed)
    ql/java/ql/test/library-tests/typeaccesses/PrintAst.qlref: Extractor error (Extraction failed)
    ql/java/ql/test/library-tests/typeaccesses/TypeAccesses.ql: Extractor error (Extraction failed)
```

Marcono1234 commented:
Thanks for the feedback and sorry for the oversights. I assume the test failures came from a compilation error in the source test file; I overlooked this and CodeCLI was not failing locally for me (see #5476).

I have addressed most of the feedback except the `wta.getBound()` check (see comment).
If you decide to merge this pull request, then please, in your own intereset, squash the commits for a clean history on `main`.

If you want I can in the future also use [fixup commits](https://thoughtbot.com/blog/autosquashing-git-commits), though I don't think GitHub automatically squashes them when merging branches.
